### ISSUE TITLE
Remove page from stack after login

### ISSUE THREAD

**Issue by eleddie** [NativeScript/nativescript-angular#309]
I know this has been asked a lot but I don't seem to find a working solution.

My app has a login page and after the user has logged in, the app goes to the main page, so far so good.
I also keep the session alive so that the user doesn't have to log in evey time the app is launched. This is done by saving a token in application settings and choose which page is going to be the default page.

After the user logs into the app, if he/she clicks the back button (I disabled the status bar in iOS to try to avoid this behavior but android still has the back button) the app goes back to the login page. I want to delete that login page from history so when the user taps the back button the application closes. Just like if I did finish() in native android activities

I've read something about backstackVisible but I don't know how to implement it with the router of angular2

Thanks in advance.

**Comment by vakrilov** [comment]
Currently, there is no way to do `backstackVisible` using the angular router. Its a feature that we hope to implement soon as the demand is high.

For now you can try one of the following approaches: 
- You can implement an authentication guard to intercept the initial navigation and redirect to login. The next time app is started after the token is read form app settings - the guard will let the navigation pass -> no login page in back-stack. There is already a branch of the groceries app that does that: check [auth.guard.ts](https://github.com/NativeScript/sample-Groceries/blob/auth-guard/app/auth.guard.ts) and [app.routes.ts](https://github.com/NativeScript/sample-Groceries/blob/auth-guard/app/app.routes.ts)
- You can use a [modal dialog](http://docs.nativescript.org/angular/ui/ui-dialogs.html#custom-dialog) to show a login dialog in your first page if login needed. The dialog will not be saved in the navigation-stack.

**Comment by vakrilov** [comment]
I have reopened the issues about `backstackVisible/clearHistory`. You can track it #284

**Comment by eleddie** [comment]
@vakrilov Thanks for your reply and for reopening the #284 issue.
I thought about executing code when the app goes back to the login page so if the user is logged in it goes back to the main page. It isn't the behavior I want but I can live with it while you are working on the new router. I saw there is the alpha version of the router 3.0.0.

The problem is that I don't know how to execute the code when the page calls, for example, the `OnResume()` method (in native android development). I am using `ngOnInit()` but it gets called only once like the constructor. Is there any way a piece of code can be executed when the page is resuming?

**Comment by vakrilov** [comment]
Just to clarify: there is already support for the `3.0.0` router in the current version(`0.1.8`). Actually my first suggestion with the `auth.guard` is based on the new routers [guards feature](https://angular.io/docs/ts/latest/guide/router.html#!#guards).

On your question - you can inject the `Page` in you component constructor (trough DI) and and attach to `navigatedTo`/`navigatingTo` events with `on()`. You can also read the `isBackNavigation` property form the [args](https://docs.nativescript.org/api-reference/interfaces/_ui_page_.navigateddata.html) of the event to apply additional logic.

**Comment by Onkarn92** [comment]
Hi @vakrilov, Is there any updates about backstackVisible/clearHistory? I'm using latest router now, but I'd like to implement clearHistory feature as i'm using login screen. Please check my [question ](https://stackoverflow.com/questions/38050881/nativescript-angular-2-use-of-back-button-click-listener)on stackoverflow at your convenience. Thanks :)

**Comment by m-abs** [comment]
@Onkarn92 ClearHistory have just been implemented:
https://github.com/NativeScript/nativescript-angular/pull/368

**Comment by vakrilov** [comment]
Yup, I have also added a [login example](https://github.com/NativeScript/nativescript-angular/blob/master/ng-sample/app/examples/router/login-test.ts) using the clear history and guards.

Closing this issue - please reopen if needed.

**Comment by sagar1911** [comment]
Hi,

How to remove a particular page from navigation ??
Let say

I am navigating to page3 from page1 via page2. when user clicks on page3 back button I don't want him to see page2 I want him to go to page1.
Is it possible in NS-Core ???
page1 -> page2 -> page3 
page3 (Back) -> page1


Thanks.

**Comment by blackrek** [comment]
> How to remove a particular page from navigation ??

when you navigate from page2 to page3 try to use router.navigate('page3url', { **replaceUrl: true** })

**Comment by sagar1911** [comment]
Hi @blackrek ,

I didn't understood router here. Here is my code, will u please tell me where to place your code snippet ??

var navigationOptions = {
            moduleName: 'views/pond/view-pond-log/view-pond-log',
            context: {
                pondid: pondidValue,
                pondName: viewModel.pondName
            },
        }
        frameModule.topmost().navigate(navigationOptions);

**Comment by vakrilov** [comment]
@sagar1911 If you are not using angular it is better to post the issue in the https://github.com/NativeScript/NativeScript repo. Here we are discussing angular and the navigation in angular apps should always be done trough the angular router abstraction ;)

**Comment by sagar1911** [comment]
okay @vakrilov ,
got it

**Comment by bhavincb** [comment]
Hi @vakrilov 

How to remove a particular pages from navigation ??
Let say

I am navigating to page2 from page1, page 3 from page2 and so on like this page1 ->page2->page3->page4->page5. 

now i want to remove page2 and page4 from stack so that when user clicks hardware back button i want navigation to behave like this page5->page3->page1->exit

**Comment by vakrilov** [comment]
Hey @bhavincb 
There is currently no direct way to modify the navigation stack. One thing we are considering is to implement navigation the [backstackVisible=false](https://docs.nativescript.org/core-concepts/navigation#navigate-without-history). However, it is not yet implemented for angular.

There are some ways to workaround this using modal views. You can check #284 where we discussed one such scenario.

**Comment by bhavincb** [comment]
hi @vakrilov ,
if there is  currently no direct way to modify the navigation stack. then please mark it as feature request/ because there are many use-cases  available where we need to modify navigation stack. this feature should be implemented when using angular.

**Comment by vakrilov** [comment]
Sure - can you log a new issue about that.

**Comment by bhavincb** [comment]
ha sure.

**Comment by bhavincb** [comment]
hii @vakrilov ,
i had created issue  [here](https://github.com/NativeScript/nativescript-angular/issues/1244).
#1244

**Comment by vishallengare61** [comment]
am new learner in angular, I use login and after login i go on home page , but home page is come on login page, how to remove login page after logged successfully and go on home page.

**Comment by vishallengare61** [comment]
I am new learner in angular, I use login and after login i go on home page , but home page is come on login page, how to remove login page after logged successfully and go on home page.
